### PR TITLE
implemented horners scheme for evaluating polynomials

### PR DESCRIPTION
**Pull request by srajangarg**
Also, there's a TODO for implementing horner's scheme for `eval_bit()` which evaluates the function at `2^x`. If that is to be done, I will have to make another recursive function just for this (because of mismatch types). So should I?

### CODE REVIEW

**Comment by Sumith1896** [comment]
Thinking over this, I think we can a another such function for `eval_bit` too. Yes, the idea will be redundant but I think that is the way to go. 
Horner's scheme has promising results in SymPy Polys.

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -159,6 +156,23 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
 }
 
+integer_class UnivariateIntPolynomial::horner_scheme(map_uint_mpz::const_iterator curr, unsigned int &l, const integer_class &x) const {
+
+    integer_class ans(1), temp;
+
+    if (curr == dict_.end())
+        return integer_class(0);
+
+    while ((*curr).first - l > 0) {
```
Comment (isuruf) [inline symengine/polynomial.cpp]: How is this while loop broken?

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -159,6 +156,23 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
+        l++;
+        ans = ans * x;
+    }
+
+    temp = (*curr).second;
+    curr++;
+    return ans * (temp + horner_scheme(curr, l, x));
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Can you turn this into a non-recursive function? For very large polynomials, recursion is not good.

File: symengine/polynomial.cpp (hunk 1 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: `l` is being incremented?

File: symengine/polynomial.cpp (hunk 2 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: I don't see a possible way, but will search into it.

**Comment by srajangarg** [comment]
The iterative solution was easier than I expected, sorry. Done it for both `eval` and `eval_bit`.

**Comment by isuruf** [comment]
Can you show me a reference to this algorithm? This appears to be different from the one in Wikipedia

File: symengine/polynomial.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/polynomial.cpp]: Ah right. I saw it as `1`

File: symengine/polynomial.cpp [hunk 3]
```diff
@@ -140,23 +140,38 @@ integer_class UnivariateIntPolynomial::max_coef() const {
...
+    integer_class ans(1), result(0);
+
     for (const auto &p : dict_) {
-        mp_pow_ui(temp, x, p.first);
-        ans += p.second * temp;
+
+        while (p.first - l > 0) {
+            l++;
+            ans = ans * x;
+        }
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Use `mp_pow_ui` to calculate `x**(p.first-l)`

**Comment by srajangarg** [comment]
Sorry @isuruf , I was implementing it from lowest degree coefficient to the highest degree originally. This used 2 multiplications and 1 addition (as opposed to 1 multiplication and 1 addition) per step. Have changed it now. Please have a look.

**Comment by isuruf** [comment]
I think lowest degree to highest is better. You can use the previous computation of `x**n` to get to `x**(n+1)`

**Comment by srajangarg** [comment]
No no, I verified online, and checked it myself. Also every article I read does it the way as shown on wikipedia. Higher to lower takes N multiplications less.

File: symengine/polynomial.cpp [hunk 4]
```diff
@@ -140,23 +140,37 @@ integer_class UnivariateIntPolynomial::max_coef() const {
...
-    for (const auto &p : dict_) {
-        mp_pow_ui(temp, x, p.first);
-        ans += p.second * temp;
+
+    unsigned int last_deg = dict_.rbegin()->first;
+    integer_class result(0), x_pow;
+
+    for (auto it = dict_.rbegin(); it != dict_.rend(); ++it) {
+
+        mp_pow_ui(x_pow, x, last_deg - (*it).first);
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Can't we use the `x_pow` from last computation?

Comment (srajangarg) [inline symengine/polynomial.cpp]: No, x_pow from the last iteration may even be bigger than needed in this iteration.
eg. `x^8 + x^2 + x`
`x_pow iteration 1 : x^6`
`x_pow iteration 2 : x`

Comment (isuruf) [inline symengine/polynomial.cpp]: You already are. Ignore my last comment

File: symengine/polynomial.cpp [hunk 5]
```diff
@@ -140,23 +140,37 @@ integer_class UnivariateIntPolynomial::max_coef() const {
...
-        ans += p.second * t;
+
+    unsigned int last_deg = dict_.rbegin()->first;
+    integer_class result(0), two_pow;
+
+    for (auto it = dict_.rbegin(); it != dict_.rend(); ++it) {
+
+        two_pow = integer_class(1) << x * (last_deg - (*it).first);
+        last_deg = (*it).first;
+        result = (*it).second + two_pow * result;
```
Comment (isuruf) [inline symengine/polynomial.cpp]: This should have used the `<<` on `result` itself.

Comment (srajangarg) [inline symengine/polynomial.cpp]: No, why?

Comment (isuruf) [inline symengine/polynomial.cpp]: Because multiplication by a power of two is slower than a bit shift

**Pull request merged**